crates/astria-conductor/src/celestia/reconstruct.rs

```diff
@@ -0,0 +1,201 @@
+
+/// Reconstructs block information from verified blocks.
+///
+/// The reconstructed blocks contain a block hash, tendermint header (both extracted from
+/// sequencer header blobs), and transactions (from rollup blobs, if present).
+///
+/// This process works in the following way:
+/// 1. Header blobs are placed in a map using a key `(block_hash, height)`.
+///    - header blobs with the same `(block_hash, height)` are dropped.
+/// 2. Each rollup blob is matched against all header blobs using Merkle proofs and roots.
+///    - A Sequencer block is reconstructed from the the first such match.
+///    - All all other sequencer header blobs with the same block hash but different height are
+///      dropped (it is assumed that sequencer blobs of different height cannot have the same block
+///      hash).
+///    - A rollup blob that has no matching sequencer header blob is dropped.
+/// 3. The remaining Sequencer blobs (i.e. those that had no matching rollup blob) are checked for
+///    `rollup_id`:
+///    - if they contained `rollup_id` they are dropped (as they should have had a matching blob but
+///      none was found).
+///    - if they did not contain `rollup_id` a Sequencer block is reconstructed.
+pub(super) fn reconstruct_blocks_from_verified_blobs(
+    verified_blobs: VerifiedBlobs,
+    rollup_id: RollupId,
+) -> Vec<ReconstructedBlock> {
+    let (celestia_height, header_blobs, rollup_blobs) = verified_blobs.into_parts();
+
+    let mut headers = BlockHashAndHeightToHeader::from_header_blobs(header_blobs);
+
+    let mut reconstructed_blocks = Vec::new();
+    for rollup in rollup_blobs {
+        if let Some(header_blob) = headers.remove_header_blob_matching_rollup_blob(&rollup) {
+            reconstructed_blocks.push(ReconstructedBlock {
+                celestia_height,
+                block_hash: header_blob.block_hash(),
+                header: header_blob.into_unchecked().header,
+                transactions: rollup.into_unchecked().transactions,
+            });
+        } else {
+            let reason = if headers.any_headers_with_block_hash(rollup.sequencer_block_hash()) {
+                "sequencer header blobs with the same block hash as the rollup blob found, but the \
+                 rollup's Merkle proof did not lead any Merkle roots"
+            } else {
+                "no sequencer header blob matching the rollup blob's block hash found"
+            };
+            info!(
+                block_hash = %base64(&rollup.sequencer_block_hash()),
+                reason,
+                "dropping rollup blob",
+            );
+        }
+    }
+
+    let remaining_headers = headers.map;
+
+    for header_blob in remaining_headers.into_values() {
+        if header_blob.contains_rollup_id(rollup_id) {
+            warn!(
+                block_hash = %base64(&header_blob.block_hash()),
+                "sequencer header blob contains the target rollup ID, but no matching rollup blob was found; the blob",
```

Comment:
also, if this case happens, we should try to fetch the corresponding rollup blob (as it should eventually be posted), is there already logic for that?

Reply:
Yeah, that's what `fetch` does.

Reply:
sorry, i meant the rollup blob might not have already been posted, but it might be posted at a later celestia block (based on the variance). is that handled here? or are we assuming the rollup blob ends up in the same block as the sequencer blob always (based on the relayer impl)?

Reply:
Right - no. We are requiring that the blobs are "atomic" in that a single sequencer block must be available in a single Celestia height.

We can probable relax this requirement and build a map of `block_hash -> rollup_blob`, but that will require a rework.

The variance refers to two subsequent sequencer heights being available within a certain window of Celestia heights.